messari-cli/src/abi.rs

```diff
@@ -0,0 +1,356 @@
+                })
+                .collect()
+        } else {
+            Vec::new()
+        };
+
+        if let Some(contract_names_str) = &self.contract_names {
+            if abi_infos.is_empty() {
+                panic!("--contract-names flag was used when the --abis flag was not! --contract-names flag can only be used when the --abis flag is used and the number of contracts \
+                entered has to match the number of abis entered.\n--contract-names=\"{}\"", contract_names_str);
+            }
+            let contract_names = contract_names_str
+                .split(",")
+                .into_iter()
+                .map(|contract_name| contract_name.to_string())
+                .collect::<Vec<_>>();
+            if contract_names.len() != abi_infos.len() {
+                panic!("Number of contract names entered does not match the number of abi entered!\n#contract-names = {}, #abis = {}", contract_names.len(), abi_infos.len());
+            }
+            (abi_infos, Some(contract_names))
+        } else {
+            (abi_infos, None)
+        }
+    }
+}
+
+pub(crate) enum AbiInfo {
+    LocalFilePath(PathBuf),
+    ContractAddress(String),
+}
+
+impl AbiInfo {
+    pub(crate) fn assert_compatible_with_protocol(&self, protocol: &Protocol) {
+        match (self, &protocol.supported_abi_addition_methods) {
+            (AbiInfo::LocalFilePath(local_file_path), SupportedAbiAdditionMethods::DownloadFromContractAddress) => panic!("Local file path: {}, supplied although protocol: {} only supports download from contract address!", local_file_path.to_string_lossy(), protocol),
+            (AbiInfo::ContractAddress(contract_address), SupportedAbiAdditionMethods::CopyFromLocalFilePath) => panic!("Contract address: {}, supplied although protocol: {} only supports copying from a local file path!", contract_address, protocol),
+            _ => {}
+        }
+    }
+}
+
+impl From<&str> for AbiInfo {
+    fn from(abi_arg: &str) -> Self {
+        if abi_arg.is_valid_abi_address() {
+            AbiInfo::ContractAddress(abi_arg.to_string())
+        } else {
+            let abi_path = Path::new(abi_arg);
+            if abi_path.exists() {
+                AbiInfo::LocalFilePath(abi_path.to_path_buf())
+            } else {
+                panic!(
+                    "Abi arg supplied: {} is neither a contract address nor a local file path!",
+                    abi_arg
+                )
+            }
+        }
+    }
+}
+
+pub(crate) fn add_abis(
+    protocol_and_network_info: ProtocolAndNetworkInfo,
+    abis_arg: &AbisArgs,
+    project_dir: &PathBuf,
+    is_add_operation: bool,
+) {
+    let (abi_infos, contract_names) = abis_arg.get_abi_info(&protocol_and_network_info.protocol);
+
+    if abi_infos.is_empty() {
+        let abi_string = if is_add_operation {
+            get_input("Abi", None, false)
+        } else {
+            let abi_string = get_input("Abi (Leave blank to skip)", Some("Abi"), true);
```

Comment:
I'd change `Abi` to `Smart Contract Address (to fetch ABI)`, which makes it more clear on what input you are asking for

Reply:
I have it as ABI because the ABI can be sourced from both smart contracts but also locally. So as it doesn't just represent a contract address I thought it would be best to keep it abi - as the umbrella term for both smart contract addresses and local file paths.

Reply:
I see. In that case, would be helpful to add some instructions around that. It wasn't clear that you could do that just by interacting with the CLI

Reply:
There are. Run any of the two commands with the --help flag and it will tell you that you can enter either a local path or a contract address